pgwire: remove readTimeoutConn in favor of a channel

rafiss opened the pull request:
Rather than using a connection that polls for the context being done every second, we now spin up an additional goroutine that blocks until the connection context is done, or the drain signal was received.

I wrote a simple benchmark of the idle connection loop to generate CPU profiles.

With the old readTimeoutConn:
<img width="861" alt="image" src="https://github.com/cockroachdb/cockroach/assets/1320573/9e0c88fc-eda7-4a51-8060-280b0d95a7a7">

With the new goroutine approach:
<img width="567" alt="image" src="https://github.com/cockroachdb/cockroach/assets/1320573/535ea962-ec03-46f7-a63b-382c65553553">

There's definitely less noise and overhead.

fixes https://github.com/cockroachdb/cockroach/issues/25585
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124373)
<!-- Reviewable:end -->

rafiss committed:
pgwire: remove readTimeoutConn in favor of a channel

Rather than using a connection that polls for the context being done
every second, we now spin up an additional goroutine that blocks until
the connection context is done, or the drain signal was received.

Release note: None

rafiss committed:
pgwire: remove unneeded test

This benchmark was only useful for capturing a CPU profile, it doesn't
measure anything else.

Release note: None

yuzefovich reviewed (commented):
Thanks for looking into this! The change overall seems good to me.

I'm not too familiar with this code, so I want to double check my understanding. IIUC currently for each connection we have two goroutines:
- the "reader" goroutine that continuously loops in `serveImpl` and see what the client is requesting us to do
- and the "processor" goroutine for the connExecutor that executes commands added into `stmtBuf` by the "reader" goroutine.

This PR introduces a third "watchdog" goroutine that monitors the state for cancellation and draining. In other words, we make a trade-off between 1) periodically forcing `net.Conn` to wake up in the "reader" due to a read deadline to see whether the "reader" should quit and 2) a separate goroutine that only monitors things with no read deadline set for the "reader". In practice, this "watchdog" goroutine should be idle most of the time, so it should remove some system calls. Does this sound right?

rafiss commented on pkg/sql/pgwire/server.go:
```diff
@@ -945,6 +942,35 @@ func (s *Server) serveImpl(
+			tcpConn, _ = underConn.(*net.TCPConn)
+		}
+		if tcpConn == nil {
+			_ = c.conn.Close()
```
hm good point. it seems safer to only close the read side.

this matches the previous behavior, where if the context was cancelled, then the readTimeoutConn would start returning errors when `Read` is called: https://github.com/cockroachdb/cockroach/blob/ceb19c5b75cf4aa02c68a669122349a3e9aca33d/pkg/sql/pgwire/server.go#L1039

that should cause the main loop of the reader goroutine to exit here: https://github.com/cockroachdb/cockroach/blob/ceb19c5b75cf4aa02c68a669122349a3e9aca33d/pkg/sql/pgwire/server.go#L1216-L1225

rafiss replied:
although, thinking about it more, the risk is that if the type switch no longer works as expected (like for example, there's a new implementation of net.Conn to deal with later), then not having a default case would make the connection hang forever.

rafiss reviewed (commented):
> This PR introduces a third "watchdog" goroutine that monitors the state for cancellation and draining. In other words, we make a trade-off between 1) periodically forcing net.Conn to wake up in the "reader" due to a read deadline to see whether the "reader" should quit and 2) a separate goroutine that only monitors things with no read deadline set for the "reader". In practice, this "watchdog" goroutine should be idle most of the time, so it should remove some system calls. Does this sound right?

yes sounds right. it's worth noting that a goroutine that does nothing apart from wait on a channel is very efficient in go. the scheduler should never try to run it.

yuzefovich reviewed (commented):
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/pgwire/server.go` line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124373#gh-1605895348:-NyM96ArEhgM8YVuM9ZG:bhs0ual) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f528eda54761c0956113c558fa96f7c60bfd12/pkg/sql/pgwire/server.go#L968)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

although, thinking about it more, the risk is that if the type switch no longer works as expected (like for example, there's a new implementation of net.Conn to deal with later), then not having a default case would make the connection hang forever.

maybe it's safer to keep the Close as a fallback? WDYT?
</blockquote></details>

Sounds reasonable to me.

I traced where `Close` is called currently, and I think it's a few layers up in the `defer` in in the async task in `netutil.TCPServer.ServeWith`, so the comment (that "serveImpl always closes the network connection before returning") is misleading before your patch but will actually become closer to reality after the patch. It might be worth adjusting the comment.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Looks like the updates haven't been pushed yet?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373#-:-NyMEcK6ABx9LjCKaw7Q:b58epck)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/pgwire/server.go` line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124373#gh-1605895348:-NyMEzZZ-4EddJRtyPPR:b-50d36w) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f528eda54761c0956113c558fa96f7c60bfd12/pkg/sql/pgwire/server.go#L968)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Sounds reasonable to me.

I traced where `Close` is called currently, and I think it's a few layers up in the `defer` in in the async task in `netutil.TCPServer.ServeWith`, so the comment (that "serveImpl always closes the network connection before returning") is misleading before your patch but will actually become closer to reality after the patch. It might be worth adjusting the comment.
</blockquote></details>

I was about to push the change, then realized that another fallback option is to call `conn.SetReadDeadline(now())`, which might be less disruptive. I'll do that instead.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124373#-:-NyMH1Yx0n5f7YhDMPpO:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/pgwire/server.go` line 968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124373#gh-1605895348:-NyMGgiDAlu4npGk_njU:b-5cxqux) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f528eda54761c0956113c558fa96f7c60bfd12/pkg/sql/pgwire/server.go#L968)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I was about to push the change, then realized that another fallback option is to call `conn.SetReadDeadline(now())`, which might be less disruptive. I'll do that instead.
</blockquote></details>

Nice, I like it.


<!-- Sent from Reviewable.io -->

rafiss commented:
tftr!

bors r=yuzefovich

rickystewart commented:
bors r=yuzefovich
